lib/autohttp/agent.js

```diff
@@ -0,0 +1,287 @@
+        path: options.socketPath
+      })
+
+      let connection
+      try {
+        const uri = options.uri
+        connection = this.http2Agent.createConnection(req, uri, http2Options)
+      } catch (e) {
+        cb(e)
+        connection && connection.socket && socketCb(connection.socket)
+        return
+      }
+
+      cb(null, 'http2', connection)
+      socketCb(connection.socket)
+
+      return
+    }
+
+    if (protocol === 'http/1.1' || protocol === 'http/1.0') {
+      const http1RequestOptions = Object.assign({}, options, {
+        agent: this.httpsAgent
+      })
+
+      let request
+      try {
+        request = https.request(http1RequestOptions)
+      } catch (e) {
+        cb(e)
+        return
+      }
+
+      request.on('socket', (socket) => socketCb(socket))
+      cb(null, 'http1', request)
+    }
+  }
+
+  createNewSocketConnection (req, options, cb, socketCb) {
+    const uri = options.uri
+    const name = this.getName(options)
+
+    const tlsSocketOptions = Object.assign({}, options, {
+      path: options.socketPath,
+      ALPNProtocols: ['h2', 'http/1.1', 'http/1.0'],
+      servername: options.servername || calculateServerName(options)
+    })
+
+    const socket = tls.connect(tlsSocketOptions)
+    socketCb(socket)
+
+    const socketConnectionErrorHandler = (e) => {
+      cb(e)
+    }
+    socket.on('error', socketConnectionErrorHandler)
+
+    socket.once('secureConnect', () => {
+      socket.removeListener('error', socketConnectionErrorHandler)
+
+      const protocol = socket.alpnProtocol
+
+      if (!protocol) {
+        cb(socket.authorizationError)
+        socket.end()
+        return
+      }
+
+      if (protocol !== 'h2' && protocol !== 'http/1.1') {
+        cb(new Error('Unknown protocol' + protocol))
+        return
+      }
+
+      // Update the cache
+      this.ALPNCache.set(name, [protocol, socket])
+
+      socket.on('close', () => {
+        // Clean the cache when the socket closes
+        this.ALPNCache.delete(name)
+      })
+
+      if (protocol === 'h2') {
+        const http2Options = Object.assign({}, options, {
+          path: options.socketPath
+        })
+        try {
+          const connection = this.http2Agent.createConnection(
+            req,
+            uri,
+            http2Options,
+            socket
+          )
+          cb(null, 'http2', connection)
+        } catch (e) {
+          cb(e)
+        }
+      }
+      if (protocol === 'http/1.1' || protocol === 'http/1.0') {
+        // Protocol is http1, using the built in agent
+        // We need to release all free sockets so that new connection is created using the overridden createconnection
+        // forcing the agent to reuse the socket used for alpn
+
+        // This reassignment works, since all code so far is sync, and happens in the same tick, hence there will be no
+        // race conditions
+        const oldCreateConnection = this.httpsAgent.createConnection
```

Comment:
Instead, why can't we extend the `https.Agent` class?

Reply:
We want to initialize the agent with any `agentoptions` that might've been passed. 
Also, by using a common agent and adding the socket to the agent when establishing a connection for the first time to the host, we can re-use the connection.